proxy.go

```diff
@@ -0,0 +1,292 @@
+		ContentLength: int64(body.Len()),
+	}
+}
+
+func (p *ProxyHTTP) log() logging.Logger {
+	if p.Log != nil {
+		return p.Log
+	}
+	return httpLog
+}
+
+// ProxyTCP is Proxy implementation for TCP streams.
+//
+// If port-based routing is used, LocalAddr or FetchLocalAddr field is required
+// for tunneling to function properly.
+// Otherwise you'll be forwarding traffic to random ports and this is usually not desired.
+//
+// If IP-based routing is used then tunnel server connection request is
+// proxied to 127.0.0.1:incomingPort where incomingPort is control message LocalPort.
+// Usually this is tunnel server's public exposed Port.
+// This behaviour can be changed by setting LocalAddr or FetchLocalAddr.
+// FetchLocalAddr takes precedence over LocalAddr.
+type ProxyTCP struct {
+	// LocalAddr defines the TCP address of the local server.
+	// This is optional if you want to specify a single TCP address.
+	LocalAddr string
+	// FetchLocalAddr is used for looking up TCP address of the server.
+	// This is optional if you want to specify a dynamic TCP address based on incommig port.
+	FetchLocalAddr func(port int) (string, error)
+	// Log is a custom logger that can be used for the proxy.
+	// If not set a "tcp" logger is used.
+	Log logging.Logger
+}
+
+// Proxy proxies remote connection to local server.
+func (p *ProxyTCP) Proxy(remote net.Conn, msg *proto.ControlMessage) {
+	if msg.Protocol != proto.TCP {
+		panic("Proxy mismatch")
+	}
+
+	var log = p.log()
+
+	var port = msg.LocalPort
+	if port == 0 {
+		log.Warning("TCP proxy to port 0")
+	}
+
+	var localAddr = fmt.Sprintf("127.0.0.1:%d", port)
+	if p.LocalAddr != "" {
+		localAddr = p.LocalAddr
+	} else if p.FetchLocalAddr != nil {
+		l, err := p.FetchLocalAddr(msg.LocalPort)
+		if err != nil {
+			log.Warning("Failed to get custom local address: %s", err)
+			return
+		}
+		localAddr = l
+	}
+
+	log.Debug("Dialing local server: %q", localAddr)
+	local, err := net.DialTimeout("tcp", localAddr, defaultTimeout)
+	if err != nil {
+		log.Error("Dialing local server %q failed: %s", localAddr, err)
+		return
+	}
+
+	Join(local, remote, log)
+}
+
+func (p *ProxyTCP) log() logging.Logger {
+	if p.Log != nil {
+		return p.Log
+	}
+	return tpcLog
+}
+
+// Join copies data between local and remote connections.
+// It reads one connection and writes to the other.
+// It aims to provide a building block for custom Proxy implementations.
+func Join(local, remote net.Conn, log logging.Logger) {
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	transfer := func(side string, dst, src net.Conn) {
+		log.Debug("proxing %s -> %s", src.RemoteAddr(), dst.RemoteAddr())
+
+		n, err := io.Copy(dst, src)
+		if err != nil {
+			log.Error("%s: copy error: %s", side, err)
+		}
+
+		if err := src.Close(); err != nil {
+			log.Debug("%s: close error: %s", side, err)
+		}
+
+		// not for yamux streams, but for client to local server connections
+		if d, ok := dst.(*net.TCPConn); ok {
+			if err := d.CloseWrite(); err != nil {
```

Comment:
Commenting for future - we should fix this method, each side of the connection should be closed like in https://github.com/nhooyr/tlswrapd/blob/9b984d7ef28e21bf6365315506494f96bad061ec/proxy.go#L69-L86.